src/client/providers/queryTransformer.ts

```diff
@@ -0,0 +1,39 @@
+import { Observable, of } from 'rxjs'
+import { flatMap, map, switchMap } from 'rxjs/operators'
+import { FeatureProviderRegistry } from './registry'
+
+export type TransformQuerySignature = (query: string) => Observable<string | null | undefined>
+
+/** Provides transformed queries from the first search extension. */
+export class QueryTransformerRegistry extends FeatureProviderRegistry<{}, TransformQuerySignature> {
+    public transformQuery(query: string): Observable<string | null | undefined> {
+        return transformQuery(this.providers, query)
+    }
+}
+
+/**
+ * Returns an observable that emits a query transformed by all providers whenever any of the last-emitted set of providers emits
+ * a query.
+ *
+ * Most callers should use QueryTransformerRegistry's transformQuery method, which uses the registered search
+ * providers.
+ */
+export function transformQuery(
+    providers: Observable<TransformQuerySignature[]>,
+    query: string
+): Observable<string | null | undefined> {
+    return providers.pipe(
+        switchMap(providers => {
+            if (providers.length === 0) {
+                return [query]
+            }
+            return providers.reduce(
+                (currentQuery, transformQuery) =>
+                    currentQuery.pipe(
+                        flatMap(q => transformQuery(q).pipe(map(transformedQuery => transformedQuery || q)))
```

Comment:
Is this the desired behavior?

- If a query transformer returns null or the empty string, use the original input string.
- Otherwise, use the original input string.

I suspect you don't intend to consider the empty string as meaning "use the original input string". But the empty string is falsey.

So I would do:

```
transformedQuery => transformedQuery === null ? q : transformedQuery
```

Or, really, why special-case null at all? If a query transformer wants to not make any modifications to the string, it can just return it unmodified. That is actually what I'd recommend.